ark/utils/notebooks_test.py

```diff
@@ -1,125 +1,561 @@
-@parametrize('pixel_cluster_col', ['pixel_meta_cluster_rename', 'pixel_som_cluster'])
-def test_cell_clustering(tb, pixel_cluster_col):
-    with tdir() as base_dir:
-        # setup the clustering process
-        fovs, chans = notebooks_test_utils.flowsom_cell_setup(
-            tb, base_dir, 'sample_pixel_dir', pixel_cluster_col=pixel_cluster_col
-        )
-
-        # mock the clustering process
-        notebooks_test_utils.flowsom_cell_cluster(
-            tb, base_dir, fovs, chans, pixel_cluster_col=pixel_cluster_col
-        )
-
-        # run the visualization and remapping process
-        notebooks_test_utils.flowsom_cell_visualize(
-            tb, base_dir, fovs, pixel_cluster_col=pixel_cluster_col
-        )
-
-
-@testbook(FIB_SEG_PATH, timeout=6000)
-def test_fib_seg(tb):
-    with tdir() as base_dir:
-        notebooks_test_utils.fib_seg_run(
-            tb, base_dir
-        )
+
+# Sets a shared notebook testing temporary directory. Saves all notebook related files in a
+# temporary directory.
+@pytest.fixture(scope="session")
+def base_dir_generator(tmp_path_factory) -> Iterator[pathlib.Path]:
+    """
+    A Fixture which creates the directory where the all notebook test inputs and outputs are
+    saved.
+    Sets the path once for all notebook related tests.
+
+    Args:
+        tmp_path_factory (pytest.TempPathFactory): Factory for temporary directories under the
+            common base temp directory.
+
+    Yields:
+        Iterator[pathlib.Path]: The iterable path containing the path of the `notebook_tests`
+        inputs and outputs.
+    """
+    notebook_test_output_dir = tmp_path_factory.mktemp("notebook_tests")
+    yield notebook_test_output_dir
+
+
+@pytest.fixture(scope="session")
+def templates_dir() -> Iterator[pathlib.Path]:
+    """
+    A Fixture which gathers the `templates` directory from `ark-analysis`.
+
+    Yields:
+        Iterator[pathlib.Path]: The directory of the templates relative to this test file.
+    """
+    templates_dir: pathlib.Path = (pathlib.Path(__file__).resolve()).parents[2] / "templates"
+    yield templates_dir
+
+
+@pytest.fixture(scope="class")
+def nb1_context(templates_dir, base_dir_generator) -> Iterator[ContextManager]:
+    """
+    Creates a testbook context manager for notebook 1.
+
+    Args:
+        templates_dir (pytest.Fixture): The fixture which yields the directory of the notebook
+            templates
+        base_dir_generator (pytest.Fixture): The fixture which yields the temporary directory
+            to store all notebook input / output.
+
+    Yields:
+        Iterator[ContextManager]: The testbook context manager which will get cleaned up
+            afterwords.
+    """
+    SEGMENT_IMAGE_DATA_PATH: pathlib.Path = templates_dir / "1_Segment_Image_Data.ipynb"
+    with testbook(SEGMENT_IMAGE_DATA_PATH, timeout=6000, execute=False) as nb_context_manager:
+        yield nb_context_manager, base_dir_generator / "nb1"
```

Comment:
This particular block is repeated several times for each notebook. Can we pass additional arguments to this function such as:

```
def nb_context(templates_dir, base_dir_generator, notebook_name, notebook_alias, notebook_abbrev) -> Iterator[ContextManager]:
    ...
    NOTEBOOK_PATH: pathlib.Path = templates_dir / notebook_name
    with testbook(NOTEBOOK_PATH, timeout=6000, execute=False) as nb_context_manager:
        yield nb_context_manager, base_dir_generator / notebook_alias
```

We can then call it accordingly.

Reply:
I don't believe so, as fixtures don't work nicely when using request to store `notebook_name` and `notebook_alias` with inherited fixtures. I tried out some stuff in [pytest-cases](https://smarie.github.io/python-pytest-cases/), but it wasn't working out.